### ISSUE TITLE
Make the optgroup labels clickable

### ISSUE THREAD

**Issue by greg0ire** [harvesthq/chosen#323]
When using a select with optgroups, having all the items in a group added to the selection when the group label is clicked would be a very handy feature. It would make selecting a group minus one item very easy too.
For instance, I have a filter form with countries grouped by zone. For the moment, if someone wants to select a group minus one item, he must select all items one by one. With this feature, he would simply click on the group and remove the country he does not want.

**Comment by tsujigiri** [comment]
A bit late to the party, but: +1 for this feature!

**Comment by KameSama** [comment]
@ruhley +1 for your solution, works like a charm!
Just a small detail that's in the full code page but not in the code segment : there's an 
else{ 
missing at the end.

**Comment by GuillaumeSTEIN** [comment]
Hi,
Thank you @ruhley for your code.
But in my case, this don't cover my needs :/
The group is not disabled after being clicked and can be reselected.

Have you done some improvements ?

**Comment by KarlBaumann** [comment]
Ist this still not implemented?

**Comment by KarlBaumann** [comment]
@GuillaumeSTEIN I tried your full version script, nothing seems to change. 
Do I need to add some special parameters or classes to HTML to make it work?
Maybe you can make a working example on http://codepen.io/ or jsfiddle?

**Comment by GuillaumeSTEIN** [comment]
Yes, you need to set option : group_selectable to true to activate this feature

**Comment by KarlBaumann** [comment]
thanks @GuillaumeSTEIN , it worked, but what I miss from @adriengibrat script, is highlighting all options under current group.

**Comment by GuillaumeSTEIN** [comment]
selected group has a "result-selected" class.
You can do some CSS to have you wished result.
Here is what I have written, you may need to adapt it for your needs :

```
 select.form-control + .chosen-container-multi .chosen-results li.result-selected{
     display: list-item;
     color: #ccc;
     cursor: default;
     background-color: white;
 }
```

**Comment by jamesalbert** [comment]
I took a stab at it #2678

**Comment by TEKFused** [comment]
This would be such a useful feature! I would love to see it implemented!

**Comment by feelslike1929** [comment]
+1 for this feature.

**Comment by Zardoz89** [comment]
If someone finds this useful, I grab @slothbear jsfiddle and improve it with : 

- "Clear all" button 
- Selecting root category item disables and deselects childrens. Why ? On my case, I would use chosen to generate a hierarchy filter, where selecting a father node implicates that would search on every children category, and we not desire to pollute the UI putting every children as selected.

http://jsfiddle.net/Zardoz/pf6dhrbc/6/

**Comment by jm93** [comment]
I just upgraded to chosen 1.8.3 in order to get the results to stay open after selecting a result on a multiple select (added in version 1.7.0).  Great feature, glad it was added. However, I had been using 1.6.1 and used the code here to allow an entire group to be added when clicking the group name: http://robido.com/jquery/add-selectdeselect-optgroup-click-events-harvests-chosen-jquery-plugin/

That code has been working great, until now. Using 1.8.3 and hide_results_on_select:false causes something to break, and it only selects the currently highlighted result option. When that line is removed (setting it back to default of true, hide upon select), it works fine and selects the entire group results options. So something in the hide_results_on_select:false is breaking/interfering with the ability to gather all the not .result-selected under a .group-result. Sacrificing this ability for the new feature is brutal.

I'll look through all the above comments to see if this has been addressed as a flaw of the new feature, but I can't tell which version people are using when they are testing this.

Any ideas?

**Comment by jm93** [comment]
okay. i've had enough. "upgrading" to 1.8.3 has been nothing but a game of whack a mole for the last week. css issues, unexpected activity issues, etc. i upgraded to get the keep results open on multi select feature (which took years to implement in the first place), but discovered it always jumps back to top when selecting from the results, so what good is that - all that waiting for nothing. performance on a single select with many options has slowed to a crawl somehow. and more issues too many to list here. good job keeping this thing going for so many years, i've enjoyed it. but the improvements came at a cost of degrading user experience and i can't have that. am researching how to pull out chosen and use select2. it's 2018, time to find a select tool that keeps up with user expectations which keep getting tougher to meet. sorry and thanks

**Comment by greg0ire** [comment]
I've moved to select2 too, but I didn't take a dump on chosen before that. You have a weird way of saying "thanks".

**Comment by yves-lndm** [comment]
Hello,
If you wish to use optgroup and be able to click on it and automatically add all the options of that group, just add this to your javascript : (the event trigger ".group-result" can be more restrictive)

I'm sure this can be added to the AbstractChosen class, with something like AbstractChosen.prototype.optgroup_click = function(evt) ....

> 

	$('body').on('click','.group-result',function(){
		// id of select
		var quest=$(this).parent().parent().parent().attr('id');
		quest='#'+quest.substring(0,quest.lastIndexOf('_'));
		// number of optgroup
		var total=$(quest+' optgroup').length;
		// number of group after group-result clicked
		var nb=0;
		$(this).nextAll().each(function(){
			if($(this).hasClass('group-result'))nb++;
		});
		// for clicked group-result, select options in the right select tag by position
		$(quest+' optgroup:nth-of-type('+(total-nb)+')').children().each(function(){
			$(this).prop('selected','selected');
		});
		// update chosen !
		$(quest).trigger('chosen:updated');
	});

**Comment by CLothering** [comment]
I made use of bootstrap-select and was also struggling to find something to be able to click on the optgroup and select all its items - This link actually worked for me, hope it helps :[https://stackoverflow.com/questions/41821115/select-deselect-optgroup-based-on-option-select-in-select-picker-boostrap](url)
